File: 76. Minimum Window Substring/minWindow.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        left = 0
        right = 0
        size = len(s)
        need = {}
        window = {}
        start = 0
        length = size + 1
        valid = 0
        for c in t:
            if need.__contains__(c):
                need[c] += 1
            else:
                need[c] = 1

        while right < size:
            curr = s[right]
            right += 1
            if need.__contains__(curr):
                if window.__contains__(curr):
                    window[curr] += 1
                else:
                    window[curr] = 1
                if need[curr] == window[curr]:
                    valid += 1 
            
            while valid == len(need):
                if right - left < length:
                    start = left
                    length = right - left
                r = s[left]
                left += 1
                if need.__contains__(r):
                    if window[r] == need[r]:
                        valid -= 1
                    window[r] -= 1
        return "" if length == size + 1 else s[start:start + length]

File: 76. Minimum Window Substring/minWindow.java
class Solution {
    public String minWindow(String s, String t) {
        int size = s.length();
        int left = 0, right = 0;
        int valid = 0;
        int start = 0, length = size + 1;
        HashMap<Character, Integer> need = new HashMap<>();
        HashMap<Character, Integer> window = new HashMap<>();
        for (int i = 0; i < t.length(); i++) {
            char temp = t.charAt(i);
            need.put(temp, need.getOrDefault(temp, 0) + 1);
        }

        while (right < size) {
            char curr = s.charAt(right);
            right++;
            if (need.containsKey(curr)) {
                window.put(curr, window.getOrDefault(curr, 0) + 1);
                if (need.get(curr).equals(window.get(curr))) {
                    valid++;
                }
            }
            
            while (valid == need.size()) {
                if (right - left < length) {
                    start = left;
                    length = right - left;
                }

                char leftChar = s.charAt(left);
                left++;
                if (need.containsKey(leftChar)) {
                    if (need.get(leftChar).equals(window.get(leftChar))) {
                        valid--;
                    }
                    window.put(leftChar, window.get(leftChar) - 1);
                }
            }
        }

        return (length == size + 1) ? "" : s.substring(start, start + length);
    }
}